main/src/cgeo/geocaching/models/CalcState.java

```diff
@@ -0,0 +1,135 @@
+    public String plainLat, plainLon;
+    public char latHemisphere, lonHemisphere;
+    public List<CalculateButton.ButtonData> buttons;
+    public List<CalculatorVariable.VariableData> equations, freeVariables;
+    public String notes; // Note, we have to use a String rather than an Editable as Editable's can't be serialized
+
+    public CalcState(final Settings.CoordInputFormatEnum format,
+                     final String plainLat,
+                     final String plainLon,
+                     final char latHem,
+                     final char lonHem,
+                     final List<CalculateButton.ButtonData> buttons,
+                     final List<CalculatorVariable.VariableData> equations,
+                     final List<CalculatorVariable.VariableData> freeVariables,
+                     final String notes) {
+        this.format = format;
+        this.plainLat = plainLat;
+        this.plainLon = plainLon;
+        latHemisphere = latHem;
+        lonHemisphere = lonHem;
+        this.buttons = buttons;
+        this.equations = equations;
+        this.freeVariables = freeVariables;
+        this.notes = notes;
+    }
+
+    public CalcState(final JSONObject json) {
+        format = Settings.CoordInputFormatEnum.values()[json.optInt("format", 2)];
+        plainLat = json.optString("plainLat");
+        plainLon = json.optString("plainLon");
+        latHemisphere = (char) json.optInt("latHemisphere", ERROR_CHAR);
+        lonHemisphere = (char) json.optInt("lonHemisphere", ERROR_CHAR);
+
+        {
+            final JSONArray array = json.optJSONArray("buttons");
+            final int arrayLength = array != null ? array.length() : 0;
+            buttons = new ArrayList<>(arrayLength);
+            for (int i = 0; i < arrayLength; i++) {
+                buttons.add(new CalculateButton.ButtonData(array.optJSONObject(i)));
+            }
+        }
+
+        {
+            final JSONArray array = json.optJSONArray("equations");
+            final int arrayLength = array != null ? array.length() : 0;
+            equations = new ArrayList<>(arrayLength);
+            for (int i = 0; i < arrayLength; i++) {
+                equations.add(new CalculatorVariable.VariableData(array.optJSONObject(i)));
+            }
+        }
+
+        {
+            final JSONArray array = json.optJSONArray("freeVariables");
+            final int arrayLength = array != null ? array.length() : 0;
+            freeVariables = new ArrayList<>(arrayLength);
+            for (int i = 0; i < arrayLength; i++) {
+                freeVariables.add(new CalculatorVariable.VariableData(array.optJSONObject(i)));
+            }
+        }
+
+        notes = json.optString("notes");
+    }
+
+    public static CalcState fromJSON(final String json) throws JSONException {
+
+        CalcState rv = null;
+        if (json != null) {
+            rv = new CalcState(new JSONObject(json));
+        }
+
+        return rv;
```

Comment:
Your proposal is even better. Early returns make it easier to understand: Here it will return, (almost) nothing will be executed after this line.

So please change this to

```java
if (json == null) {
    return null;
}

return new CalcState(new JSONObject(json));
```